design pages: Enable passkey authentication

ikerexxe opened the pull request:
passkey - Authentication in a centralized environment

Relates to https://github.com/SSSD/sssd/issues/6228

abbra commented:
In the design page there is no explanation how registration process happens and what parameters used. It took me some time to find out that a relying party, for example, is the name of the SSSD domain user belongs to. I think these details should be part of the documentation.

ikerexxe commented:
> In the design page there is no explanation how registration process happens and what parameters used. It took me some time to find out that a relying party, for example, is the name of the SSSD domain user belongs to. I think these details should be part of the documentation.

I've added a new section, `Registration process`, to explain it.

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,313 @@
+------------------------
+The objective is to use a passkey to locally authenticate a user against 
+centralized identity management system. For that purpose an integration with an
+LDAP server like AD or FreeIPA is needed. SSSD should get the user credentials
```
s/SSSD/SSSD clients/

Also, I would explicitly state that first the passkey data needs to be added to the LDAP directory server.

ikerexxe replied:
I did both things

justin-stephenson replied:
Thanks, LGTM.

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,313 @@
+
+The implementation will loosely follow existing smartcard authentication in
+SSSD. On top of that, and as mentioned before, the information will be stored
+in two places: in the LDAP server and in the passkey. The server will contain
```
s/in the passkey/on the passkey/  sounds a bit better IMO

ikerexxe replied:
Yeah, this is a recurrent problem for Spanish-speaking, we fail when using in/on/at. Thanks for catching it.

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,313 @@
+In the Active Directory case, the
+`altSecurityIdentities <https://learn.microsoft.com/en-us/windows/win32/adschema/a-altsecurityidentities>`__
+LDAP attribute will be used to store the passkey mapping data. Add this to
+``src/proviers/ad/ad_opts.c`` to retrieve from the AD user object. Example:
```
Probably this line is not needed about `src/providers/ad`ad_opts.c`

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,313 @@
+  verification during authentication. The default is that the key itself
+  decides what to do.
+
+* debug-libfido2: flag to print libfido2 library messages.
```
defaults to false

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,313 @@
+An example prompt interaction can look like: ::
+
+    Insert your Passkey device, then press ENTER. 
+    ENTER   
```
<ENTER> to signify it is not text but a keyboard 'enter'  being pressed

ikerexxe commented:
@justin-stephenson I fixed the comments and I added `If the credential is discoverable, then the passkey_child will also print the userId, and the PAM responder has to compare it with the one provided by the LDAP server. If they match then it can return PAM_SUCCESS.` in the `Call the passkey child in the PAM responder` section.

andreboscatto commented:
@abbra it would be nice if you could take a look :)

abbra commented:
@andreboscatto I will, I need to get through this flu first...

pbrezina commented:
Hi Iker, can you please squash the commits together? It will make it easier to provide comments.

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+command line the system administrator would use 
+`sssctl <https://docs.pagure.org/sssd.sssd/design_pages/sssctl.html>`__. This
+would enable to unify the user experience. Then, the output of this command can
+be set to the LDAP server in a dedicated attribute. For AD we could use the
```
is the attribute choice decided already for each ID provider? If so, please replace 'we could use' by a more definitive statement that lists the attributes per each ID provider.

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+
+IPA process
++++++++++++
+FreeIPA would provide a direct way to register the passkey attributes to the
```
I think for IPA you should provide a link to readthedocs page. Since IPA part is not merged yet (chicken and egg problem with SSSD release with passkey yet), we should probably update this part of the design once it is published. It is definitely good to replace 'would provide' and below 'can be done' by more concrete statement based on the IPA design.  Please see IPA design here: https://github.com/flo-renaud/freeipa/blob/passkey/doc/designs/passkeys.md

ikerexxe replied:
I'm keeping this open until until IPA part is merged.

ikerexxe replied:
Opened https://github.com/SSSD/sssd.io/pull/69 to include the link

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+
+LDAP attributes
+***************
+Newly added IPA ``ipapasskey`` LDAP attribute will be added to
```
Please rephrase as factual statement. E.g. 'FreeIPA ID provider will use LDAP attribute ``ipaPassKey`` to retrieve the passkey mapping data ...".

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+
+* altSecurityIdentities: passkey:9S87qLk8/RxYJ3skwwYduomAM+/HDtz41N0+w/vRL6aGKJkLMsg+2OhO0E8pK5DuO1KmdK61K8PmH7jiYuOqbg==,9YE1s/f7J47h2A/DXCVFWulqoBXFzCSxcbGEBadkpSUFjwUudhPLnPUTv2qNamakXJgRYCZQ7vpS/t5zXMLnkw==
+
+Finally, newly added IPA ``iparequireuserverification`` LDAP attribute will be
```
Same here. Please rephrase -- no need for 'newly added ... LDAP attribute', it will not be new by the time it is merged in IPA already, just state the attribute name.

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+Passkey - Authentication in a centralized environment
```
A general observation. Now that we have most components implemented, it would make sense to rephrase those parts of the design that speak in potential way to state facts. E.g. instead of "will use X" or "would use X", just state the choice, "implementation uses X".

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+
+User authentication with the key
+********************************
+The idea is to create a helper process similar to the p11_child. The helper
```
Please rephrase -- we already have implementation, this is not an idea anymore.

abbra commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+(passkey_verification, passkey_child_timeout, debug_libfido2) and set up the
+arguments (Domain, Key-handle and Public-key) for the passkey-child.
+
+Execute the passkey_child passing the arguments listed above. The passkey PIN
```
It is unclear under what user the passkey_child is launched. Since passkeys often require access to USB or other external bus to communicate, it would be good to mention how to achieve that (udev rules). Specifically, this should be covered if passkey_child does not run as root.

ikerexxe replied:
I explain this at the top of the document, where I explain how to use the `sssctl` command for the first time.

abbra reviewed (changes requested):
Hi @ikerexxe, thank you for the design document. In general, it is good one already. I just left few comments that ask for changes since we already have made certain changes since this document was written and it would be better to simply state the decisions rather than list the choices that weren't yet finalized when the document was written.

ikerexxe commented:
@abbra thanks for the feedback. As a general note, I have rephrased all potential statements (will/would) to affirmative phrases as requested.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+---------
+As a user, I want to walk to any system on a network and log in (desktop and/or
+terminal) using a passkey device connected to the system. The user credentials
+will be stored in a central management system like FreeIPA or Active Directory.
```
ATM We only support local login over SSSD+FIDO. We don't support SSH+SSSD+FIDO, right? Therefore "I want to walk to any system on a network" is not part of this design page.

ikerexxe replied:
Right, we only support SSSD+FIDO. The part of the phrase that you are quoting is exactly the one that specifies that you need to be physically near the system to perform the authentication, you can't do it remotely by SSH or any other means.

pbrezina replied:
Please, rephrase it to make it more clear. To me, "walk to any system on a network" means to ssh to any system on a network.

abbra replied:
I think you should drop 'any system on a network'  and replace it with something like 'a system enrolled into a centralized identity management solution', this would be more clear.

ikerexxe replied:
I rephrased it to take into account your feedback

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+Another difference is that with the current IPA feature, the first and second
+factor can also be used for doing a direct LDAP bind. This will not be possible
+with the new passkey scheme. Direct LDAP bind support will not be in the scope
+of this implementation.
```
I found this paragraph confusing. Is direct bind with FIDO supported on IPA but unsupported on plain LDAP? Or do you refer to OTP?

abbra replied:
I think this paragraph can be dropped, it does not add anything to clarify the implementation.

ikerexxe replied:
Direct bind isn't supported in the scope of this feature. In this section of the document I'm explaining the differences between the already implement OTP feature and this one.

It's kind of confusing because the FIDO2 keys provided by several vendors also contain other protocols (OTP, PIV, etc.). So, even if we refer to them as FIDO2 keys, they provide FIDO2 and other mechanisms. When I started writing this document and asked for feedback, some people didn't understand why the actual implementation (OTP) wasn't good enough. Our users were already using FIDO2 keys to authenticate, why did we need another implementation? That's why I wrote this section.

abbra replied:
I think what is needed here is an explicit statement that only FIDO2 part is supported here. If a hardware token implements other authentication mechanisms aside from FIDO2, these aren't considered by the passkey authentication feature. What is done by the FreeIPA server side should not be documented here.

ikerexxe replied:
I dropped this section from the document as it doesn't make sense in this context and I added an explanation in the `Problem statement` to indicate that his feature is FIDO2 only.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+use the passkey attribute.
+
+The format for
+`server-side credentials <https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Resident_Keys.html>`__
```
Are you sure this link is valid? It is the same as the one for discoverable credentials and does not explain server-side creds.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+`altSecurityIdentities <https://learn.microsoft.com/en-us/windows/win32/adschema/a-altsecurityidentities>`__
+with a prefix (i.e. passkey). For FreeIPA we use a new attribute called
+ipapasskey from the ipapasskeyuser objectclass. For any other LDAP server we
+use the passkey attribute.
```
Provide a link to the passkey ldap schema?

ikerexxe replied:
Yes, that sounds good to me. I'll keep this open until the FreeIPA changes are merged in the "central" repository and then I'll add the link.

ikerexxe replied:
Opened https://github.com/SSSD/sssd.io/pull/69 to include the link

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+entry. As specified in FreeIPA's documentation this can be done with the
+`user-add-passkey` command or directly in the WebUI. This way FreeIPA can take
+advantage of the SSSD and libfido2 integration to communicate with the
+device in a single step.
```
So you either call sssctl + user-add-passkey, or you use webui which talks to the fido2 device directly? Or how does it work?

ikerexxe replied:
It's `ipa user-add-passkey` or WebUI. I've added the `ipa` command to make it more clear.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+++++++++++++++++++++++++++++++++++++++
+The users, or the system administrator, calls sssctl. The user can manually
+add this information to their AD object using AD's 'User and Computers'
+utility, or any other attribute in case of another LDAP server.
```
The administrator must have temporary access to the fido2 device, right?

ikerexxe replied:
Yes

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+result and call ``pam_reply()``. If the credential is discoverable, then the
+passkey_child also prints the ``userId``, and the PAM responder has to
+compare it with the one provided by the LDAP server. If they match, then it can
+return PAM_SUCCESS.
```
Comparing of userId might be better to do in passkey_child so the whole logic in in single process.

ikerexxe replied:
@sumit-bose I remember you asked to do the comparison in the PAM responder, but I can't remember why. Can you enlighten us?

sumit-bose replied:
Hi,

the idea was, since the userID is created as a random number, ot handle it as a second shared secret so that e.g. in the FreeIPA Kerberos case the userID is only know to the server and the token. The userID will not be send to the client by the server but only returned fom the client to the server so that the server can check if userID is the same as the stored one. Similar for the local authentication, passkey_child should not know which value is expected here.

Originally I thought that the userId will be part of the authenticator data returned by the token and hence is protected against modification by the signature. But it looks it can only be send separately. Not sure if it would need some additional protection in this case?

bye,
Sumit

ikerexxe replied:
As discussed in the meeting this part of the design is fine. So I'm resolving this conversation.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,318 @@
+
+The most basic example of a registration would be the following: ::
+
+    # sssctl passkey-exec --register --username=USERNAME --domain=DOMAIN
```
I know I'm late and I haven't reviewed patches that already land in, but I'm not very fond of current implementation.

```
errno_t sssctl_passkey_exec(struct sss_cmdline *cmdline,
                            struct sss_tool_ctx *tool_ctx,
                            void *pvt)
{
    errno_t ret;

    ret = sssctl_wrap_command(SSS_PASSKEY_CHILD, cmdline, tool_ctx, pvt);

    return ret;
}
```

Wouldn't it be better to provide `sssctl passkey-register` command instead of allowing to call anything in passkey_child?

ikerexxe replied:
That makes sense. @justin-stephenson what do you think?

pbrezina replied:
But users need to call only register, right? You can call passkey_child directly for debugging and testing purpose.

ikerexxe replied:
Opened https://issues.redhat.com/browse/SSSD-5644 to track this work.

pbrezina reviewed (commented):
Hi Iker, see few comments inside.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,295 @@
+  verification during authentication. The default is that the key itself
+  decides what to do.
+
+* debug-libfido2: flag to print libfido2 library messages. Defaults to false.
```
Should this really be inside passkey_verification option?  Wouldn't a standalone option be better?

ikerexxe replied:
I'd like to keep the configuration behaviour as similar to the `p11_child` as possible. Changing it to a stand-alone option would do the contrary.

pbrezina replied:
I'm sorry, but this is debugging, it does not have anything to do with verification so it does not belong there. What similarities are you referring to? I don't see any debug value in `certificate_verification`.

ikerexxe replied:
I meant, that all the configuration parameters for `p11_child` are defined in `certificate_verification`, and I'd like to do the same for passkey. It's true though that this is a debug parameter. Are you suggesting to create a new passkey section and move it to its own configuration option?

pbrezina replied:
You can create passkey_debug_libfido2 inside [pam].

ikerexxe replied:
Done. Justin will take of changing the implementation to match this change.

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,295 @@
+* interactive: set to prompt a message and wait before testing the presence of
+  a passkey device. Recommended if your device doesn’t have a tactile trigger.
+
+* inter-prompt=your prompt here: set individual prompt message for interactive
```
Underscore instead of dash?

ikerexxe replied:
Right, and it's `interactive_prompt` instead of `inter_prompt` so I've also updated that

pbrezina commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,295 @@
+
+* touch: set to prompt a message to remind the user to touch the device.
+
+* touch-prompt=your prompt here: set individual prompt message for the cue
```
Underscore instead of dash?

ikerexxe replied:
Fixed

justin-stephenson commented on src/design-pages/passkey_authentication.rst:
```diff
@@ -0,0 +1,295 @@
+"passkey_child_timeout" sets the timeout for the PAM responder to wait for
+passkey_child to finish.
+
+"debug_libfido2" prints libfido2 library messages. Defaults to false.
```
This will need to be updated to indicate it is now under the [pam] section, and I renamed it to `passkey_debug_libfido2`

ikerexxe replied:
Done, and thanks for taking care of it!

pbrezina reviewed (approved):
Thank you, ack.

abbra commented:
LGTM. The Github UI is not allowing to add approval.

URLs should be updated once IPA merged the PR for passkeys which will only happen after SSSD merged both phases.
